File: python/CardGameEx/Ex05.py
# Ex05.py

from ch.aplu.jgamegrid import Location
from ch.aplu.jcardgame import Deck, CardGame, CardCover, RowLayout
from ch.aplu.jcardgame.Hand import SortType
from enum import enum

# ------------------ global functions --------------------------------------
def dealingOut():
   global hands
   global talon
   cg.setStatusText("Dealing out in batches...")
   hands = deck.dealingOut(nbPlayers, 0)  # All cards in shuffled talon
   talon = hands[nbPlayers]

   for i in range(nbPlayers):
      hands[i].setView(cg, RowLayout(handLocations[i], 300, 90 * i))
      hands[i].draw()

   talonCover = CardCover(cg, talonLocation, deck, 1, 0)
  
   while not talon.isEmpty() and not cg.isDisposed():
      for i in range(nbPlayers):
         for k in range(packetSize):
            hands[i].insert(talon.getLast(), False)
            talon.removeLast(False)
         cardCover = CardCover(cg, talonLocation, deck, 1, 0)
         cardCover.slideToTarget(handLocations[i], 10, True, True)
         cardCover.removeSelf()
         if i == 0:
            hands[i].setVerso(False)
            hands[i].sort(SortType.RANKPRIORITY, False)
         else:
            hands[i].setVerso(True)
         hands[i].draw()
   
   talonCover.removeSelf()
   hands[0].setTouchEnabled(True)

# ------------------------ main --------------------------------------------
Suit = enum("SPADES", "HEARTS", "DIAMONDS", "CLUBS")
Rank = enum("ACE", "KING", "QUEEN", "JACK", "TEN", "NINE", "EIGHT", "SEVEN", "SIX")
deck = Deck(Suit.values(), Rank.values(), "cover");

hands = None
talon = None
nbPlayers = 4
handLocations = [Location(300, 525), Location(75, 300), Location(300, 75), Location(525, 300)]
talonLocation = Location(300, 300)
packetSize = 3
  
cg = CardGame(600, 600, 30)
dealingOut()
cg.setStatusText("Dealing Out...done")

File: cardgame_tutorial/Ex05.java
// Ex05.java

import ch.aplu.jcardgame.*;
import ch.aplu.jgamegrid.*;

public class Ex05 extends CardGame
{
  public enum Suit
  {
    SPADES, HEARTS, DIAMONDS, CLUBS
  }

  public enum Rank
  {
    ACE, KING, QUEEN, JACK, TEN, NINE, EIGHT, SEVEN, SIX
  }
  //

  private final Deck deck =
    new Deck(Suit.values(), Rank.values(), "cover");
  private final int nbPlayers = 4;
  private final Location[] handLocations =
  {
    new Location(300, 525),
    new Location(75, 300),
    new Location(300, 75),
    new Location(525, 300),
  };
  private final Location talonLocation = new Location(300, 300);
  private final int packetSize = 3;
  private Hand[] hands;
  private Hand talon;

  public Ex05()
  {
    super(600, 600, 30);
    dealingOut();
    setStatusText("Dealing out...done.");
  }

  private void dealingOut()
  {
    setStatusText("Dealing out in batches...");

    hands = deck.dealingOut(nbPlayers, 0);  // All cards in shuffled talon
    talon = hands[nbPlayers];

    for (int i = 0; i < nbPlayers; i++)
    {
      hands[i].setView(this, new RowLayout(handLocations[i], 300, 90 * i));
      hands[i].draw();
    }

    CardCover talonCover = new CardCover(this, talonLocation, deck, 1, 0);

    while (!talon.isEmpty())
    {
      for (int i = 0; i < nbPlayers; i++)
      {
        for (int k = 0; k < packetSize; k++)
        {
          hands[i].insert(talon.getLast(), false);
          talon.removeLast(false);
        }
        CardCover cardCover = new CardCover(this, talonLocation, deck, 1, 0);
        cardCover.slideToTarget(handLocations[i], 10, true, true);
        cardCover.removeSelf();
        if (i == 0)
        {
          hands[i].setVerso(false);
          hands[i].sort(Hand.SortType.RANKPRIORITY, false);
        }
        else
          hands[i].setVerso(true);
        hands[i].draw();
      }
    }
    talonCover.removeSelf();
    hands[0].setTouchEnabled(true);
  }

  public static void main(String[] args)
  {
    new Ex05();
  }

}
